Change horizontal pod autoscaler to use otelcol scale subresource

kevinearls opened the pull request:
Signed-off-by: Kevin Earls <kearls@redhat.com>

This fixes #801 so that the HPA is created on an OpenTelemetryCollector rather than a Deployment.

I believe it also fixes #922 
I believe it also fixes #906 but will confirm

pavolloffay commented on apis/v1alpha1/opentelemetrycollector_types.go:
```diff
@@ -37,6 +37,10 @@ type OpenTelemetryCollectorSpec struct {
 
+	// MinReplicas sets a lower bound to the autoscaling feature
+	// +optional
+	MinReplicas int32 `json:"minReplicas,omitempty"`
```
What is the relationship between `MinReplicas` and `Replicas`? Which one should be used when? This is something that should be documented.

pavolloffay replied:
With that we probably need to add some checks to the webhook 

https://github.com/open-telemetry/opentelemetry-operator/blob/b77c89d6ea846ba5d184a4ab8716744e190ef677/apis/v1alpha1/opentelemetrycollector_webhook.go#L120-L122

to test the minreplicas or any relationship between replicas, minreplias and maxreplicas.

pavolloffay commented on pkg/collector/horizontalpodautoscaler.go:
```diff
@@ -35,18 +35,18 @@ func HorizontalPodAutoscaler(cfg config.Config, logger logr.Logger, otelcol v1al
 	return autoscalingv1.HorizontalPodAutoscaler{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:        naming.Collector(otelcol),
+			Name:        otelcol.Name,
```
What is the rationale for the name change? Could this produce an invalid name?

kevinearls replied:
We are now scaling on the OpenTelemetryCollector rather than the deployment.  If the otelcol.Name is "simple" we append "-collector" to the deployment name.  

If you think it's a good idea I could add a function to naming which just does

```
	return DNSName(Truncate("%s", 63, otelcol.Name))
```

pavolloffay commented:
I would like to see e2e test for this functionality. It can be actually two tests -  scaling up and down with the HPA.

yuriolisa reviewed (changes requested):
Please fix the lint.

kevinearls commented:
@pavolloffay Please review.  I think I've covered all of the things we discussed.

@yuriolisa lint now passes

pavolloffay commented on apis/v1alpha1/opentelemetrycollector_types.go:
```diff
@@ -37,9 +48,8 @@ type OpenTelemetryCollectorSpec struct {
 
-	// MaxReplicas sets an upper bound to the autoscaling feature. If MaxReplicas is set autoscaling is enabled.
 	// +optional
-	MaxReplicas *int32 `json:"maxReplicas,omitempty"`
```
I am not if we can change the CRD. In this case a field is removed.

pavolloffay commented on controllers/opentelemetrycollector_controller.go:
```diff
@@ -21,7 +21,8 @@ import (
 	"github.com/go-logr/logr"
 	appsv1 "k8s.io/api/apps/v1"
-	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	autoscalingv2beta2 "k8s.io/api/autoscaling/v2beta2"
```
I would prefer to keep using v1 if the v2beta2 is not needed.

kevinearls commented:
@pavolloffay Can you please review this, re-running the checks first if needed?  (The 1.24 e2e tests are failing in a place I didn't touch.)

I have reverted to using autoscaling v1 and removed the AutoScaleSpec changes (I can do both later) so I think that covers everything we've discussed.

kevinearls commented on hack/install-metrics-server.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+# Install metrics-server for autoscale tests
```
No, with minikube you can just add --addons "metrics-server" to the start command.

Where should I document this?  In this script?  In the Makefile?

yuriolisa reviewed (approved):
@kevinearls, please ensure that change will not disrupt the existing instances like @pavolloffay mentioned above. LGTM.

kevinearls commented:
@yuriolisa @pavolloffay I have added code that takes care of the upgrade.  Please review and merge if it looks ok.

yuriolisa commented on apis/v1alpha1/opentelemetrycollector_webhook.go:
```diff
@@ -120,6 +120,15 @@ func (r *OpenTelemetryCollector) validateCRDSpec() error {
+		}
+
+		if r.Spec.MinReplicas != nil && *r.Spec.MinReplicas < int32(1) {
+			return fmt.Errorf("the OpenTelemetry Spec autoscale configuration is incorrect, minReplicas should be more than one")
```
I consider MinReplicas should be at least one, correct?

yuriolisa commented on bundle/manifests/opentelemetry.io_opentelemetrycollectors.yaml:
```diff
@@ -223,6 +223,10 @@ spec:
                 format: int32
                 type: integer
+              minReplicas:
+                description: MinReplicas sets a lower bound to the autoscaling feature.
```
Could you please add here the lowest value?

yuriolisa commented on docs/api.md:
```diff
@@ -1455,6 +1455,15 @@ OpenTelemetryCollectorSpec defines the desired state of OpenTelemetryCollector.
+        <td><b>minReplicas</b></td>
+        <td>integer</td>
+        <td>
+          MinReplicas sets a lower bound to the autoscaling feature.<br/>
```
Also here.

yuriolisa reviewed (approved):
LGTM.

kevinearls commented:
Closing this in favor of #983 as I managed to get myself into rebase hell.
